### PR TITLE
[RFC] C++ `=default` ctors/dtors and the use of non-static member initializer syntax.

### PR DESCRIPTION
**Pull request by christianparpart**
Interesting references in the CppCoreGuidelines

* [C.45 Don’t define a default constructor that only initializes data members; use in-class member initializers instead](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c45-dont-define-a-default-constructor-that-only-initializes-data-members-use-in-class-member-initializers-instead)
* [C.48: Prefer in-class initializers to member initializers in constructors for constant initializers](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c48-prefer-in-class-initializers-to-member-initializers-in-constructors-for-constant-initializers)

### CODE REVIEW

File: libdevcore/IndentedWriter.h [hunk 1]
```diff
@@ -61,7 +61,7 @@ class IndentedWriter
 		unsigned indentation;
 	};
 
-	std::vector<Line> m_lines;
+	std::vector<Line> m_lines = {{std::string(), 0}};
```
Comment (chriseth) [inline libdevcore/IndentedWriter.h]: Let's see if that works with old compilers...

File: libevmasm/Assembly.h [hunk 2]
```diff
@@ -45,7 +45,7 @@ using AssemblyPointer = std::shared_ptr<Assembly>;
 class Assembly
 {
 public:
-	Assembly() {}
```
Comment (chriseth) [inline libevmasm/Assembly.h:48]: Wouldn't it be better to actually remove it?

File: libevmasm/KnownState.cpp [hunk 3]
```diff
@@ -304,7 +304,7 @@ KnownState::StoreOperation KnownState::storeInStorage(
 
 	AssemblyItem item(Instruction::SSTORE, _location);
 	Id id = m_expressionClasses->find(item, {_slot, _value}, true, m_sequenceNumber);
-	StoreOperation operation(StoreOperation::Storage, _slot, m_sequenceNumber, id);
+	StoreOperation operation {StoreOperation::Storage, _slot, m_sequenceNumber, id};
```
Comment (chriseth) [inline libevmasm/KnownState.cpp]: No space in front of `{`

File: liblangutil/SourceReferenceExtractor.h [hunk 4]
```diff
@@ -31,28 +31,27 @@ namespace langutil
...
 struct SourceReference
 {
-	std::string message;    ///< A message that relates to this source reference (such as a warning or an error message).
-	std::string sourceName; ///< Underlying source name (for example the filename).
-	LineColumn position;    ///< Actual (error) position this source reference is surrounding.
-	bool multiline;         ///< Indicates whether the actual SourceReference is truncated to one line.
-	std::string text;       ///< Extracted source code text (potentially truncated if multiline or too long).
-	int startColumn;        ///< Highlighting range-start of text field.
-	int endColumn;          ///< Highlighting range-end of text field.
+	std::string message {};    ///< A message that relates to this source reference (such as a warning or an error message).
```
Comment (ekpyron) [inline liblangutil/SourceReferenceExtractor.h]: We still need to decide whether we want to default-member-initialize non-PODs.
It's somewhat more consistent to generally member-initialize everything under all circumstances, but on the other hand it's exactly the same as not member-initializing those members like that.
I'm fine with both anyways.

File: libsolidity/parsing/Parser.h [hunk 5]
```diff
@@ -48,6 +48,9 @@ class Parser: public langutil::ParserBase
 	struct VarDeclParserOptions
 	{
 		VarDeclParserOptions() {}
+		// This is actually not needed, but due to a defect in the C++ standard, we have to.
```
Comment (ekpyron) [inline libsolidity/parsing/Parser.h:50]: Maybe move this before the constructor and start with ``The following constructor is actually...``?

**Review by ekpyron — COMMENTED** [review]
I like this, but the style changes are still needed.

File: libdevcore/IndentedWriter.h [hunk 6]
```diff
@@ -34,7 +34,7 @@ DEV_SIMPLE_EXCEPTION(IndentedWriterError);
 class IndentedWriter
 {
 public:
-	explicit IndentedWriter(): m_lines(std::vector<Line>{{std::string(), 0}}) {}
+	IndentedWriter() = default;
```
Comment (chriseth) [inline libdevcore/IndentedWriter.h]: I think if `ClassName() = default;` is the only constructor, we should rather remove it completely.

File: liblangutil/CharStream.h [hunk 7]
```diff
@@ -92,9 +92,9 @@ class CharStream
 	///@}
 
 private:
-	std::string m_source;
-	std::string m_name;
-	size_t m_position;
+	std::string m_source {};
```
Comment (chriseth) [inline liblangutil/CharStream.h]: In my opinion, this is unnecessary clutter and reduces readability. You have the impression that something special is going on while it's not.

File: libevmasm/KnownState.cpp (hunk 3 above)
Comment (christianparpart) [reply libevmasm/KnownState.cpp]: Actually, I personally prefer spacing here, for readability. Also goes in line with `inline class member initialization`, such as `int m_value {1234};`. it would just look unreadable in not having that space there (but that's my opinion). What do you think? Maybe we should globally revisit our coding style guidelines in general (as a team). :)

File: liblangutil/SourceReferenceExtractor.h (hunk 4 above)
Comment (christianparpart) [reply liblangutil/SourceReferenceExtractor.h]: @ekpyron Actually it does make sense the way I wrote it, here, because it eliminates the need of constructors as long as you don't need custom logic in the constructor bodies.

Comment (ekpyron) [reply liblangutil/SourceReferenceExtractor.h]: Are you sure? I'm pretty certain that it'd be *exactly* the same, if you would just use ``std::string message;`` - in both cases it will default construct even without defining a constructor. You only need to member-initialize for PODs or types that are not default-constructible.

Comment (ekpyron) [reply liblangutil/SourceReferenceExtractor.h]: Or if you don't want the member to be default-constructed of course :-).

File: libevmasm/KnownState.h [hunk 8]
```diff
@@ -74,18 +74,13 @@ class KnownState
...
+
 		bool isValid() const { return target != Invalid; }
-		Target target;
-		Id slot;
-		unsigned sequenceNumber;
-		Id expression;
+
+		Target target {Invalid};
+		Id slot {0};
+		unsigned sequenceNumber {-1u};
```
Comment (ekpyron) [inline libevmasm/KnownState.h]: It seems like MSVC doesn't like ``-1u`` after all (it warns and has ``Werr`` and therefore fails)... the cleanest way to avoid that would probably be ``std::numeric_limits<unsigned>::max()`` (although ``-1u`` is actually standard conformant, so MSVC is a bit overzealous in warning about it here).

File: libdevcore/IndentedWriter.h [hunk 9]
```diff
@@ -61,7 +59,7 @@ class IndentedWriter
 		unsigned indentation;
 	};
 
-	std::vector<Line> m_lines;
+	std::vector<Line> m_lines {{std::string(), 0}};
```
Comment (chriseth) [inline libdevcore/IndentedWriter.h]:
```suggestion
	std::vector<Line> m_lines{{std::string(), 0}};
```
(why should the spacing be different than for statements?)

File: libevmasm/KnownState.h [hunk 10]
```diff
@@ -74,18 +74,13 @@ class KnownState
...
-		): target(_target), slot(_slot), sequenceNumber(_sequenceNumber), expression(_expression) {}
+
 		bool isValid() const { return target != Invalid; }
-		Target target;
-		Id slot;
-		unsigned sequenceNumber;
-		Id expression;
+
+		Target target {Invalid};
+		Id slot {0};
```
Comment (chriseth) [inline libevmasm/KnownState.h]: This was uninitialized before???

Comment (chriseth) [reply libevmasm/KnownState.h]: I would also set this to max.

File: libevmasm/KnownState.h [hunk 11]
```diff
@@ -74,18 +74,13 @@ class KnownState
...
+		unsigned sequenceNumber {std::numeric_limits<unsigned>::max()};
+		Id expression {0};
```
Comment (chriseth) [inline libevmasm/KnownState.h]: same here

File: liblangutil/EVMVersion.h [hunk 12]
```diff
@@ -39,7 +39,7 @@ class EVMVersion:
 	boost::equality_comparable<EVMVersion>
 {
 public:
-	EVMVersion() {}
+	EVMVersion() = default;
```
Comment (chriseth) [inline liblangutil/EVMVersion.h:42]: Why not remove?

File: liblangutil/SourceReferenceExtractor.h [hunk 13]
```diff
@@ -31,28 +31,27 @@ namespace langutil
...
 {
-	std::string message;    ///< A message that relates to this source reference (such as a warning or an error message).
-	std::string sourceName; ///< Underlying source name (for example the filename).
-	LineColumn position;    ///< Actual (error) position this source reference is surrounding.
-	bool multiline;         ///< Indicates whether the actual SourceReference is truncated to one line.
-	std::string text;       ///< Extracted source code text (potentially truncated if multiline or too long).
-	int startColumn;        ///< Highlighting range-start of text field.
-	int endColumn;          ///< Highlighting range-end of text field.
+	std::string message;       ///< A message that relates to this source reference (such as a warning or an error message).
+	std::string sourceName {}; ///< Underlying source name (for example the filename).
```
Comment (chriseth) [inline liblangutil/SourceReferenceExtractor.h]:
```suggestion
	std::string sourceName;    ///< Underlying source name (for example the filename).
```

File: libsolidity/ast/Types.h [hunk 14]
```diff
@@ -601,7 +601,7 @@ class FixedBytesType: public Type
 class BoolType: public Type
 {
 public:
-	BoolType() {}
+	BoolType() = default;
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: Is this needed?

File: solc/CommandLineInterface.h [hunk 15]
```diff
@@ -41,7 +41,7 @@ enum class DocumentationType: uint8_t;
 class CommandLineInterface
 {
 public:
-	CommandLineInterface() {}
+	CommandLineInterface() = default;
```
Comment (chriseth) [inline solc/CommandLineInterface.h]: Needed?

File: test/libsolidity/GasMeter.cpp [hunk 16]
```diff
@@ -44,7 +44,7 @@ namespace test
 class GasMeterTestFramework: public SolidityExecutionFramework
 {
 public:
-	GasMeterTestFramework() { }
+	GasMeterTestFramework() = default;
```
Comment (chriseth) [inline test/libsolidity/GasMeter.cpp]: Needed?

File: test/libsolidity/SolidityABIJSON.cpp [hunk 17]
```diff
@@ -59,7 +59,7 @@ class JSONInterfaceChecker
 	}
 
 protected:
-	CompilerStack m_compilerStack;
+	CompilerStack m_compilerStack {};
```
Comment (chriseth) [inline test/libsolidity/SolidityABIJSON.cpp]:
```suggestion
	CompilerStack m_compilerStack;
```

File: test/libsolidity/SolidityABIJSON.cpp [hunk 18]
```diff
@@ -38,7 +38,7 @@ namespace test
 class JSONInterfaceChecker
 {
 public:
-	JSONInterfaceChecker(): m_compilerStack() {}
```
Comment (chriseth) [inline test/libsolidity/SolidityABIJSON.cpp:41]: Needed?

File: test/libsolidity/SolidityNatspecJSON.cpp [hunk 19]
```diff
@@ -77,7 +77,7 @@ class DocumentationChecker
 	}
 
 private:
-	CompilerStack m_compilerStack;
+	CompilerStack m_compilerStack {};
```
Comment (chriseth) [inline test/libsolidity/SolidityNatspecJSON.cpp]:
```suggestion
	CompilerStack m_compilerStack;
```

File: test/libsolidity/SolidityNatspecJSON.cpp [hunk 20]
```diff
@@ -39,7 +39,7 @@ namespace test
 class DocumentationChecker
 {
 public:
-	DocumentationChecker(): m_compilerStack() {}
```
Comment (chriseth) [inline test/libsolidity/SolidityNatspecJSON.cpp:42]: Needed?

File: test/libsolidity/SolidityOptimizer.cpp [hunk 21]
```diff
@@ -45,7 +45,7 @@ namespace test
 class OptimizerTestFramework: public SolidityExecutionFramework
 {
 public:
-	OptimizerTestFramework() { }
```
Comment (chriseth) [inline test/libsolidity/SolidityOptimizer.cpp:48]: Needed?

**Comment by chriseth** [comment]
Build error on windows: C:\projects\solidity\liblangutil\ParserBase.cpp(103): error C2440: 'initializing': cannot convert from 'initializer list' to 'langutil::SourceLocation'

File: liblangutil/EVMVersion.h (hunk 12 above)
Comment (christianparpart) [reply liblangutil/EVMVersion.h:42]: it's actually needed to enable default initialization of private members. (since all member variables are private, we have to have a default constructor in order to be able to write something like `EVMVersion version;` (otherwise the caller-side can't initialize.)

File: liblangutil/SourceReferenceExtractor.h (hunk 13 above)
Comment (christianparpart) [reply liblangutil/SourceReferenceExtractor.h]: @chriseth ths initializer can't be fully omitted. It serves like a default value if none is given when constructing this object.

Comment (ekpyron) [reply liblangutil/SourceReferenceExtractor.h]: @christianparpart I'm still pretty sure, you're wrong about that - it will default construct even without this default value - just the same as ``message`` one line above, where you *did* remove the member-initialization already :-).
Non-PODs that have a default constructor will always be default constructed using their default-constructor, regardless of whether they have an empty member-initializer or not.

**Comment by christianparpart** [comment]
> Build error on windows: C:\projects\solidity\liblangutil\ParserBase.cpp(103): error C2440: 'initializing': cannot convert from 'initializer list' to 'langutil::SourceLocation'

We need to decide whether we want to continue to support VS 2015 it seems. It's supporting almost nothing from C++14 (as we found out in #solidity-dev Gitter channel, and most likely, the only user may be AppVeyor).

File: liblangutil/EVMVersion.h (hunk 12 above)
Comment (ekpyron) [reply liblangutil/EVMVersion.h:42]: @christianparpart I'm pretty sure that's not true. Try this, it'll work:

```
#include <iostream>
class A {
public:
    A() { std::cout << "A::A()" << std::endl; }
};
class B {
private:
    A a;
};

int main()
{
    B b;
}
```

Comment (ekpyron) [reply liblangutil/EVMVersion.h:42]: For reference see ``Implicitly-declared default constructor`` (https://en.cppreference.com/w/cpp/language/default_constructor)

File: liblangutil/SourceReferenceExtractor.h (hunk 13 above)
Comment (christianparpart) [reply liblangutil/SourceReferenceExtractor.h]: @ekpyron the way the code was, with the above proposal applied, you run into a compile error (at least on my compiler, GCC 8.1), so I empirically fixed that until GCC was happy. Now, we decided to go with another approach anyways.

File: liblangutil/SourceLocation.h [hunk 22]
```diff
@@ -53,9 +49,9 @@ struct SourceLocation
 
 	bool isEmpty() const { return start == -1 && end == -1; }
 
-	int start;
-	int end;
-	std::shared_ptr<CharStream> source;
+	int start {-1};
+	int end {-1};
+	std::shared_ptr<CharStream> source {};
```
Comment (chriseth) [inline liblangutil/SourceLocation.h]:
```suggestion
	std::shared_ptr<CharStream> source;
```

File: liblangutil/SourceLocation.h [hunk 23]
```diff
@@ -53,9 +49,9 @@ struct SourceLocation
 
 	bool isEmpty() const { return start == -1 && end == -1; }
 
-	int start;
-	int end;
-	std::shared_ptr<CharStream> source;
+	int start {-1};
```
Comment (chriseth) [inline liblangutil/SourceLocation.h]:
```suggestion
	int start{-1};
```
Or could we perhaps just use `= -1` for some here? It's so much more readable.
What about `int start = {-1}`?

File: libyul/YulString.h [hunk 24]
```diff
@@ -81,8 +80,8 @@ class YulStringRepository: boost::noncopyable
 	}
 	static constexpr std::uint64_t emptyHash() { return 14695981039346656037u; }
 private:
-	std::vector<std::shared_ptr<std::string>> m_strings;
-	std::unordered_multimap<std::uint64_t, size_t> m_hashToID;
+	std::vector<std::shared_ptr<std::string>> m_strings {std::make_shared<std::string>()};
+	std::unordered_multimap<std::uint64_t, size_t> m_hashToID {{emptyHash(), 0}};
```
Comment (chriseth) [inline libyul/YulString.h]: Actually it's really hard to follow what happens here. Perhaps we should even make this more verbose and create a constructor
```
YulStringRepository() { insert(string()); }
```
and create an `insert` function that handles most of `stringToHandle` except for the empty check?

**Comment by chriseth** [comment]
Could we split this into parts that do not need any discussion and parts that do need discussion?

File: libyul/YulString.h (hunk 24 above)
Comment (ekpyron) [reply libyul/YulString.h]: Do you really think that initializing a map is *that* hard to read? Especially if we move it to a member-initializer here I think it's quite readable actually... Or do you mean why this and ``m_strings`` is initialized that way? If that's the case I think a comment might help more than creating another function...

File: liblangutil/SourceReferenceExtractor.h (hunk 13 above)
Comment (ekpyron) [reply liblangutil/SourceReferenceExtractor.h]: What kind of error? The code should be exactly equivalent, so this would rather indicate a bug in GCC 8.1 then... In any case, if it's now fine, it doesn't matter too much.

File: libyul/YulString.h (hunk 24 above)
Comment (chriseth) [reply libyul/YulString.h]: As soon as I see `{{`, I start wondering about how deeply nested everything is. We should perhaps not discuss this on this PR, but I think adding some code that clarifies that this initialization satisfy the invariant about the empty string would help readability.

File: liblangutil/SourceReferenceExtractor.h (hunk 13 above)
Comment (chriseth) [reply liblangutil/SourceReferenceExtractor.h]: @ekpyron if everything apart from a single member has an initializer, the compiler will generate a constructor that only requires you to specify a single value.

Comment (ekpyron) [reply liblangutil/SourceReferenceExtractor.h]: @chriseth I highly doubt that and in fact it doesn't, if I try it. The only constructors that the compiler ever generates implicitly are default, copy and move constructors. The only thing that would work is aggregate initialization but for that I can skip values (at the end) anyways, regardless of whether they are member-initialized or not. (As a side note: pre-C++14 member-initialization would have prevented aggregate initialization - so nice to have C++14 :-)).

File: libsolidity/parsing/Parser.h [hunk 25]
```diff
@@ -99,7 +102,7 @@ class Parser: public langutil::ParserBase
 	ASTPointer<FunctionTypeName> parseFunctionType();
 	ASTPointer<Mapping> parseMapping();
 	ASTPointer<ParameterList> parseParameterList(
-		VarDeclParserOptions const& _options,
```
Comment (chriseth) [inline libsolidity/parsing/Parser.h:102]: Was not there before.

Comment (chriseth) [reply libsolidity/parsing/Parser.h:102]: I think this is still mergeable like that.

**Review by chriseth — APPROVED** [review]
